discovery_engine_core/core/src/message_api.rs

```diff
@@ -0,0 +1,232 @@
+//!   is done separately from this interface and their FFI bindings
+//!     - depending on the FFI bindings we either send a package consisting
+//!       of a bytes-message and a magic cookie, or we append the magic cookie
+//!       to the byte message
+//!
+//! In the future we can split this module and language bindings non specific
+//! to this module out of this repo as a set of libraries if we thing it's helpful.
+//!
+//! Normally we would make `Engine` implement `System`, but for the beginning this
+//! has to exist in parallel with the previous FFI API. As such we have to to temp.
+//! do some things differently:
+//!
+//! 1. implement System in the binding crate, instead of the core crate where it is
+//!    supposed to be implemented
+//! 2. do some hacky thing to make sure both the FFI and message API have access to
+//!    the engine
+//!
+
+use std::{
+    any::Any,
+    panic::AssertUnwindSafe,
+    sync::{
+        atomic::{self, AtomicI64},
+        Arc,
+    },
+};
+
+use async_trait::async_trait;
+use futures::FutureExt;
+use thiserror::Error;
+
+/// Initialize the system, setting up message passing.
+///
+/// # Panic
+///
+/// - This panics if it is not called in the context of a tokio
+///   runtime, the message handling loop will run in the given
+///   tokio runtime.
+pub fn init<S, F, T>(init_message: Package, send_response: F, recv_request: T)
+where
+    S: System<F>,
+    F: SendResponse,
+    T: RecvRequest,
+{
+    tokio::spawn(async move {
+        let send_last_response = send_response.clone();
+        // All parts moved into this future are not used after a unwind catching,
+        // similar the future is not resumed after a unwind catching.
+        // Side Note: "unwind safety" has nothing to do with rust safety.
+        let run_engine_fut = AssertUnwindSafe(async move {
+            let cookie = init_message.cookie;
+            let mut system = match S::create(init_message, send_response.clone()).await {
+                Ok(system) => system,
+                Err(err) => {
+                    return Some(Package {
+                        cookie,
+                        message: err,
+                    });
+                }
+            };
+
+            while let Some(msg) = recv_request.recv().await {
+                let cookie = msg.cookie;
+                let message = system
+                    .handle_message(msg)
+                    .await
+                    .unwrap_or_else(|err_msg| err_msg);
+
+                if let Err(ResponseChannelBroken) =
+                    send_response.send(Package { message, cookie }).await
+                {
+                    return None;
+                }
+            }
+
+            let message = system.shutdown().await;
+            Some(Package {
+                message,
+                cookie: MagicCookie::generate_cookie(),
+            })
+        });
+
+        let res = run_engine_fut
+            .catch_unwind()
```

Comment:
This line won't compile in this scatch, basically assert unwind safe/catch unwind are fundamentally broken with futures in most situations :cry: